pkg/util/clusterdata/clusterdata.go

```diff
@@ -0,0 +1,129 @@
+import (
+	"context"
+	"sync"
+
+	azureproviderapis "github.com/openshift/cluster-api-provider-azure/pkg/apis"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = serializer.NewCodecFactory(scheme)
+
+func init() {
+	utilruntime.Must(azureproviderapis.AddToScheme(scheme))
+}
+
+// OpenShiftClusterEnricher must update the cluster object with
+// data received from API server
+type OpenShiftClusterEnricher interface {
+	Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster) error
+}
+
+type enricherFeatureConstructor func(*rest.Config, *api.OpenShiftCluster) (enricherFeature, error)
+type enricherFeature interface {
+	Name() string
+	FetchData(ctx context.Context) (func(), error)
+}
+
+// NewBestEffortEnricher returns an enricher that attempts to populate
+// fields, but ignores errors in case of failures
+func NewBestEffortEnricher(log *logrus.Entry, env env.Interface) OpenShiftClusterEnricher {
+	return &bestEffortEnricher{
+		log: log,
+		env: env,
+
+		restConfig: restconfig.RestConfig,
+		featureConstructors: []enricherFeatureConstructor{
+			newClusterVersionEnricherFeature,
+			newServicePrincipalEnricherFeature,
+			newWorkerProfilesEnricherFeature,
+		},
+	}
+}
+
+type bestEffortEnricher struct {
+	log *logrus.Entry
+	env env.Interface
+
+	restConfig          func(env env.Interface, oc *api.OpenShiftCluster) (*rest.Config, error)
+	featureConstructors []enricherFeatureConstructor
+}
+
+func (e *bestEffortEnricher) Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster) error {
+	// TODO: Make use of the deadline from ctx
+	// NOTE: The above might need to wait for this KEP to land in openshift/client-go:
+	//       https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20200123-client-go-ctx.md
+
+	var wg sync.WaitGroup
+	wg.Add(len(ocs))
+	errs := make([]error, len(ocs), len(ocs))
+	for i := range ocs {
+		i := i // https://golang.org/doc/faq#closures_and_goroutines
+		go func() {
+			defer wg.Done()
+			errs[i] = e.enrichOne(ctx, ocs[i])
+		}()
+	}
+	wg.Wait()
+
+	for i := range errs {
+		if errs[i] != nil {
+			e.log.Errorf("error %q while processing cluster %q", errs[i].Error(), ocs[i].ID)
+		}
+	}
+
+	return nil
+}
+
+func (e *bestEffortEnricher) enrichOne(ctx context.Context, oc *api.OpenShiftCluster) error {
+	restConfig, err := e.restConfig(e.env, oc)
+	if err != nil {
+		return err
+	}
+
+	features := make([]enricherFeature, len(e.featureConstructors), len(e.featureConstructors))
+	for i := range e.featureConstructors {
+		feature, err := e.featureConstructors[i](restConfig, oc)
+		if err != nil {
+			return err
+		}
+		features[i] = feature
+	}
+
+	var wg sync.WaitGroup
```

Comment:
`f(oc)` will be blocking, no? So If `ctx` has 10s timeot and first `f(oc)` takes 40s we will cancel other functions only after 40s. Or do I miss something?

Reply:
I mean that f(oc) is just the closure function which quickly writes the fetched value to oc.

Reply:
I believe we still want to run all callbacks even if timeout expired. In case of expired timeout we will be applying callbacks with default values.

Reply:
For history: we will be applying default values earlier, on instantiation an `f` will be applying values received from the API server.

Also to clarify the need for `select`: it will help us prevent cases when `FetchData` gets stuck for whatever reason (context doesn't get used or bug in code or something else).